Add GCC 12.5 as the alternative compiler

sparlane opened the pull request:
Brings in GCC 12.5 to replace 4.7 as the alternative compiler.

Includes patches to enable -Werror for both world and kernel.

sparlane committed:
Add contrib/gcc-12

sparlane committed:
Add makefiles for gcc120

Based on the makefiles for gcc 8.3, by way of 9.5, 10.5, and 11.5

gcc has added:
* a new library "libcody"
* a new place for compiler-related tools "c++tools"
* the static analyzer "gcc/analyzer"

Also, all (C++) .c files in gcc/ have been renamed to .cc

sparlane committed:
gcc120: apply changes from gcc80

bd23261bc0 libstdc++: Fix unsigned wraparound in codecvt::do_length [PR105857]
fca8e87b2e gcc: Define ElfW() if it is not defined.
2007c584a7 gcc8: Fix the fix.
9cf0c62db7 gccs had ElfW() already. Comment out for now.
2ecd6c6623 gcc: Allow our C++ headers to survive -Wsystem-headers (2).
894c19b2d8 gcc: Allow our C++ headers to survive -Wsystem-headers.
de2610f57c Adjust makefiles for GCC80 update.
df642abc04 gcc80: Add local -save-temps=objects option.
a69def156b gcc80: Handle TZ specific "%+" format in strftime.
fcad8200ea gcc80: Remove throw() from protos in libstdc++ headers.
39964fd19e gcc80: Include <stdlib.h> for alloca() prototype.
3dcb3c95ea gcc80: Add alternative variant of pthread presence checking.
632577dbc0 gcc80: Adjust search paths for base compiler layout.
ed36d35d08 gcc80: Adjust default dynamic linker search location.

sparlane committed:
Copy attributes for strong/weak reference

GCC 9 and later take issue with mismatched attributes between aliased functions

This is based on the change made in linux for the same issue

sparlane committed:
cc120: Work around needing to build omp-device-properties.h

The makefile rule I did have didn't trigger:
make[4]: don't know how to make omp-device-properties.h. Stop

This file is always going to be the same, so just include the generated one.

sparlane committed:
Disable narrowing warning when gcc10+ is used

sparlane committed:
Import gcc 12.5.0

Taken from c17d40bb3778bca5e81595f033df9222b66658eb (releases/gcc-12.5.0)

Filtered with:
 # for f in `cat README.DELETED`; do rm -fr $f ; done

sparlane committed:
Make gcc 120 the alternative compiler

sparlane committed:
Remove gcc47 support

sparlane committed:
Enable Werror when building kernel with gcc 12.5

Disables the new warnings that have appeared from newer GCC versions

sparlane committed:
gcc120: remove d language support

sparlane committed:
Remove gcc47 related files

sparlane committed:
Enable Werror for gcc120

sparlane commented:
@liweitianux I haven't yet made 2 changes that you have previously mentioned:
1) Swap to using the man pages from the source, primarily because the git repo doesn't actually contain these, the ones I have added when adding the makefiles are from the gcc-12.5.0.tar.xz tarball.
2) Dropped the version number from the makefile name conditionals.

liweitianux commented:
>     1. Swap to using the man pages from the source, primarily because the git repo doesn't actually contain these, the ones I have added when adding the makefiles are from the gcc-12.5.0.tar.xz tarball.

The current `contrib/gcc-8.0` have the whole `gcc/doc` directory removed, so they're missing. I suggest we keep the wanted `.1` man pages there and then we can directly install them from our makefiles.

>     2. Dropped the version number from the makefile name conditionals.

I suggest we do this if there is no conflict.

sparlane commented:
1) I imported the code from gcc's git repo which doesn't have the .1 files, https://gcc.gnu.org/git/?p=gcc.git;a=tree;f=gcc/doc;h=590ea95318855317081c003c1ad06393bbc73d86;hb=c17d40bb3778bca5e81595f033df9222b66658eb

2) is done now, didn't seem to cause any issues.

sparlane committed:
cc120: Fix -save-temp=objects

An upstream patch considerably reworked how the -save-temps option (and friends) works
https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=1dedc12d186a110854537e1279b4e6c29f2df35a

Adjust the patch we have to compensation for these changes

sparlane commented:
Actually I've put back the version numbers in makefile.inc's based on what swildner said in IRC.

sparlane committed:
Add manpages from gcc tarball

These are only in the tarball and not present in the repo

sparlane committed:
Merge branch 'gcc12' into use_gcc125

liweitianux commented on contrib/gcc-12/gcc/gcc.cc:
```diff
@@ -277,8 +277,9 @@ static enum save_temps {
   SAVE_TEMPS_CWD,		/* -save-temps in current directory */
   SAVE_TEMPS_DUMP,              /* -save-temps in dumpdir */
   SAVE_TEMPS_OBJ,		/* -save-temps in object directory */
```
Please remove the `,` to reduce an unnecessary diff.

sparlane commented on contrib/gcc-12/gcc/gcc.cc:
```diff
@@ -4450,7 +4451,10 @@ driver_handle_option (struct gcc_options *opts,
+	{
+	  keep_suffix = true;
+          save_temps_flag = SAVE_TEMPS_OBJ;
+	}
```
Oops, not sure how I got that commit id wrong, updated.
Probably should use a text editor that shows tabs/spaces for this.

liweitianux commented:
I've finally integrated this work, except for the warning flags and `-Werror` enabling. We'll have to fix the meaningful warnings caught by GCC 12 and then tweak the flags to disable the false ones.

Thank you for the hard work, @sparlane!
